airbyte-db/src/main/java/io/airbyte/db/JdbcUtils.java

```diff
@@ -0,0 +1,231 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.db;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.airbyte.commons.functional.CheckedFunction;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.protocol.models.Field.JsonSchemaPrimitive;
+import java.math.BigDecimal;
+import java.sql.JDBCType;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class JdbcUtils {
+
+  private static final DateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'"); // Quoted "Z" to indicate UTC, no timezone offset
+
+  /**
+   * Map records returned in a result set.
+   *
+   * @param resultSet the result set
+   * @param mapper function to make each record of the result set
+   * @param <T> type that each record will be mapped to
+   * @return stream of records that the result set is mapped to.
+   */
+  public static <T> Stream<T> mapResultSet(ResultSet resultSet, CheckedFunction<ResultSet, T, SQLException> mapper) {
+    return StreamSupport.stream(new Spliterators.AbstractSpliterator<T>(Long.MAX_VALUE, Spliterator.ORDERED) {
+
+      @Override
+      public boolean tryAdvance(Consumer<? super T> action) {
+        try {
+          if (!resultSet.next())
+            return false;
+          action.accept(mapper.apply(resultSet));
+          return true;
+        } catch (SQLException e) {
+          throw new RuntimeException(e);
+        }
+      }
+
+    }, false);
+  }
+
+  /**
+   * Collect each record of a ResultSet into a list of JsonNode.
+   *
+   * @param resultSet the result set
+   * @return List of JsonNode.
+   * @throws SQLException exceptions throws when parsing the ResultSet.
+   */
+  public static List<JsonNode> resultSetToJson(ResultSet resultSet) throws SQLException {
```

Comment:
Same comment as above. someone is going to mess it up. Can you just make it return the stream<JsonNode> instead?

Reply:
Could you help me understand _what_ someone is going to mess up? 

Given the interface is outputting a list, it seems to be say loud and clear that "this better all be able to fit into memory if you're going to use it." But may I am misunderstanding the thing that you think people will mess up.